File: apps/go/simulator/chemistry.go
package simulator

import (
	"math"
	"math/rand"
	"sync/atomic"
)

// moleculeIDGen is the global molecule ID counter.
var moleculeIDGen atomic.Int64

// Molecule represents a collection of bonded atoms.
type Molecule struct {
	Atoms            []*Atom
	Name             string
	Formula          string
	ID               int64
	Energy           float64
	Position         [3]float64
	IsOrganic        bool
	FunctionalGroups []string
}

// NewMolecule creates a new molecule.
func NewMolecule(atoms []*Atom, name string) *Molecule {
	m := &Molecule{
		Atoms:            atoms,
		Name:             name,
		ID:               moleculeIDGen.Add(1),
		FunctionalGroups: make([]string, 0),
	}
	if len(atoms) > 0 {
		m.Position = atoms[0].Position
	}
	m.computeFormula()
	return m
}

func (m *Molecule) computeFormula() {
	counts := make(map[string]int)
	for _, a := range m.Atoms {
		counts[a.Symbol()]++
	}

	hasC := false
	hasH := false
	for _, a := range m.Atoms {
		if a.AtomicNumber == 6 {
			hasC = true
		}
		if a.AtomicNumber == 1 {
			hasH = true
		}
	}
	m.IsOrganic = hasC && hasH

	if m.Formula != "" {
		return
	}

	formula := ""
	// Standard chemistry ordering: C, H, then alphabetical
	for _, sym := range []string{"C", "H"} {
		if n, ok := counts[sym]; ok {
			if n > 1 {
				formula += sym + string(rune('0'+n)) // simplified for small counts
			} else {
				formula += sym
			}
			delete(counts, sym)
		}
	}
	// Remaining alphabetical
	for sym, n := range counts {
		if n > 1 {
			formula += sym + string(rune('0'+n))
		} else {
			formula += sym
		}
	}
	m.Formula = formula
}

// MolecularWeight returns the sum of atom mass numbers.
func (m *Molecule) MolecularWeight() int {
	w := 0
	for _, a := range m.Atoms {
		w += a.MassNumber
	}
	return w
}

// ChemicalSystem manages molecular assembly and chemical reactions.
type ChemicalSystem struct {
	Atomic           *AtomicSystem
	Molecules        []*Molecule
	ReactionsOccurred int
	WaterCount       int
	AminoAcidCount   int
	NucleotideCount  int
	rng              *rand.Rand
}

// NewChemicalSystem creates a new chemical system.
func NewChemicalSystem(atomicSystem *AtomicSystem, rng *rand.Rand) *ChemicalSystem {
	return &ChemicalSystem{
		Atomic:    atomicSystem,
		Molecules: make([]*Molecule, 0, 64),
		rng:       rng,
	}
}

// unbondedAtoms returns atoms of a given atomic number that have no bonds (or fewer than maxBonds).
func (cs *ChemicalSystem) unbondedAtoms(atomicNumber int, maxBonds int) []*Atom {
	var result []*Atom
	for _, a := range cs.Atomic.Atoms {
		if a.AtomicNumber == atomicNumber && len(a.Bonds) < maxBonds {
			result = append(result, a)
		}
	}
	return result
}

// FormWater forms water molecules: 2H + O -> H2O.
func (cs *ChemicalSystem) FormWater() []*Molecule {
	var waters []*Molecule
	hydrogens := cs.unbondedAtoms(1, 1)  // H with 0 bonds
	oxygens := cs.unbondedAtoms(8, 2)     // O with < 2 bonds

	for len(hydrogens) >= 2 && len(oxygens) > 0 {
		h1 := hydrogens[len(hydrogens)-1]
		hydrogens = hydrogens[:len(hydrogens)-1]
		h2 := hydrogens[len(hydrogens)-1]
		hydrogens = hydrogens[:len(hydrogens)-1]
		o := oxygens[len(oxygens)-1]
		oxygens = oxygens[:len(oxygens)-1]

		water := NewMolecule([]*Atom{h1, h2, o}, "water")
		water.Position = o.Position
		waters = append(waters, water)
		cs.Molecules = append(cs.Molecules, water)
		cs.WaterCount++

		h1.Bonds = append(h1.Bonds, o.ID)
		h2.Bonds = append(h2.Bonds, o.ID)
		o.Bonds = append(o.Bonds, h1.ID, h2.ID)
	}

	return waters
}

// FormMethane forms methane: C + 4H -> CH4.
func (cs *ChemicalSystem) FormMethane() []*Molecule {
	var methanes []*Molecule
	carbons := cs.unbondedAtoms(6, 1)
	hydrogens := cs.unbondedAtoms(1, 1)

	for len(carbons) > 0 && len(hydrogens) >= 4 {
		c := carbons[len(carbons)-1]
		carbons = carbons[:len(carbons)-1]

		hs := make([]*Atom, 4)
		for i := 0; i < 4; i++ {
			hs[i] = hydrogens[len(hydrogens)-1]
			hydrogens = hydrogens[:len(hydrogens)-1]
		}

		atoms := append([]*Atom{c}, hs...)
		methane := NewMolecule(atoms, "methane")
		methane.Position = c.Position
		methanes = append(methanes, methane)
		cs.Molecules = append(cs.Molecules, methane)

		for _, h := range hs {
			c.Bonds = append(c.Bonds, h.ID)
			h.Bonds = append(h.Bonds, c.ID)
		}
	}

	return methanes
}

// FormAmmonia forms ammonia: N + 3H -> NH3.
func (cs *ChemicalSystem) FormAmmonia() []*Molecule {
	var ammonias []*Molecule
	nitrogens := cs.unbondedAtoms(7, 1)
	hydrogens := cs.unbondedAtoms(1, 1)

	for len(nitrogens) > 0 && len(hydrogens) >= 3 {
		n := nitrogens[len(nitrogens)-1]
		nitrogens = nitrogens[:len(nitrogens)-1]

		hs := make([]*Atom, 3)
		for i := 0; i < 3; i++ {
			hs[i] = hydrogens[len(hydrogens)-1]
			hydrogens = hydrogens[:len(hydrogens)-1]
		}

		atoms := append([]*Atom{n}, hs...)
		ammonia := NewMolecule(atoms, "ammonia")
		ammonia.Position = n.Position
		ammonias = append(ammonias, ammonia)
		cs.Molecules = append(cs.Molecules, ammonia)

		for _, h := range hs {
			n.Bonds = append(n.Bonds, h.ID)
			h.Bonds = append(h.Bonds, n.ID)
		}
	}

	return ammonias
}

// FormAminoAcid forms an amino acid from available atoms (simplified).
// Requires 2C + 5H + 2O + 1N minimum (glycine).
func (cs *ChemicalSystem) FormAminoAcid(aaType string) *Molecule {
	carbons := cs.unbondedAtoms(6, 1)
	hydrogens := cs.unbondedAtoms(1, 1)
	oxygens := cs.unbondedAtoms(8, 2)
	nitrogens := cs.unbondedAtoms(7, 1)

	if len(carbons) < 2 || len(hydrogens) < 5 || len(oxygens) < 2 || len(nitrogens) < 1 {
		return nil
	}

	atoms := make([]*Atom, 0, 10)
	for i := 0; i < 2; i++ {
		atoms = append(atoms, carbons[len(carbons)-1])
		carbons = carbons[:len(carbons)-1]
	}
	for i := 0; i < 5; i++ {
		atoms = append(atoms, hydrogens[len(hydrogens)-1])
		hydrogens = hydrogens[:len(hydrogens)-1]
	}
	for i := 0; i < 2; i++ {
		atoms = append(atoms, oxygens[len(oxygens)-1])
		oxygens = oxygens[:len(oxygens)-1]
	}
	atoms = append(atoms, nitrogens[len(nitrogens)-1])

	aa := NewMolecule(atoms, aaType)
	aa.Position = atoms[0].Position
	aa.IsOrganic = true
	aa.FunctionalGroups = []string{"amino", "carboxyl"}
	cs.Molecules = append(cs.Molecules, aa)
	cs.AminoAcidCount++

	// Form internal bonds
	for i := 1; i < len(atoms); i++ {
		atoms[0].Bonds = append(atoms[0].Bonds, atoms[i].ID)
		atoms[i].Bonds = append(atoms[i].Bonds, atoms[0].ID)
	}

	return aa
}

// FormNucleotide forms a nucleotide (sugar + phosphate + base).
// Simplified: requires C5 + H8 + O4 + N2 minimum.
func (cs *ChemicalSystem) FormNucleotide(base string) *Molecule {
	carbons := cs.unbondedAtoms(6, 1)
	hydrogens := cs.unbondedAtoms(1, 1)
	oxygens := cs.unbondedAtoms(8, 2)
	nitrogens := cs.unbondedAtoms(7, 1)

	if len(carbons) < 5 || len(hydrogens) < 8 || len(oxygens) < 4 || len(nitrogens) < 2 {
		return nil
	}

	atoms := make([]*Atom, 0, 19)
	for i := 0; i < 5; i++ {
		atoms = append(atoms, carbons[len(carbons)-1])
		carbons = carbons[:len(carbons)-1]
	}
	for i := 0; i < 8; i++ {
		atoms = append(atoms, hydrogens[len(hydrogens)-1])
		hydrogens = hydrogens[:len(hydrogens)-1]
	}
	for i := 0; i < 4; i++ {
		atoms = append(atoms, oxygens[len(oxygens)-1])
		oxygens = oxygens[:len(oxygens)-1]
	}
	for i := 0; i < 2; i++ {
		atoms = append(atoms, nitrogens[len(nitrogens)-1])
		nitrogens = nitrogens[:len(nitrogens)-1]
	}

	nuc := NewMolecule(atoms, "nucleotide-"+base)
	nuc.Position = atoms[0].Position
	nuc.IsOrganic = true
	nuc.FunctionalGroups = []string{"sugar", "phosphate", "base"}
	cs.Molecules = append(cs.Molecules, nuc)
	cs.NucleotideCount++

	for i := 1; i < len(atoms); i++ {
		atoms[0].Bonds = append(atoms[0].Bonds, atoms[i].ID)
		atoms[i].Bonds = append(atoms[i].Bonds, atoms[0].ID)
	}

	return nuc
}

// CatalyzedReaction runs catalyzed reactions to form complex molecules.
func (cs *ChemicalSystem) CatalyzedReaction(temperature float64, catalystPresent bool) int {
	formed := 0
	eaFactor := 1.0
	if catalystPresent {
		eaFactor = 0.3
	}

	thermal := KB * temperature

	// Try to form amino acids
	if thermal > 0 && len(cs.Atomic.Atoms) > 10 {
		aaProb := math.Exp(-5.0 * eaFactor / (thermal + 1e-20))
		if cs.rng.Float64() < aaProb {
			aaType := AminoAcids[cs.rng.Intn(len(AminoAcids))]
			if cs.FormAminoAcid(aaType) != nil {
				formed++
				cs.ReactionsOccurred++
			}
		}
	}

	// Try to form nucleotides
	if thermal > 0 && len(cs.Atomic.Atoms) > 19 {
		nucProb := math.Exp(-8.0 * eaFactor / (thermal + 1e-20))
		if cs.rng.Float64() < nucProb {
			bases := []string{"A", "T", "G", "C"}
			base := bases[cs.rng.Intn(len(bases))]
			if cs.FormNucleotide(base) != nil {
				formed++
				cs.ReactionsOccurred++
			}
		}
	}

	return formed
}

// MoleculeCensus returns molecule counts by type name.
func (cs *ChemicalSystem) MoleculeCensus() map[string]int {
	counts := make(map[string]int)
	for _, m := range cs.Molecules {
		key := m.Name
		if key == "" {
			key = m.Formula
		}
		counts[key]++
	}
	return counts
}

File: apps/rust/src/simulator/chemistry.rs
//! Chemistry simulation -- molecular assembly and reactions.
//!
//! Models formation of molecules from atoms: water, amino acids,
//! nucleotides, and other biomolecules essential for life.
//! Chemical reactions are driven by energy, catalysis, and concentration.

use rand::Rng;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};

use super::atomic::{Atom, AtomicSystem};
use super::constants::*;

static MOLECULE_ID_COUNTER: AtomicU64 = AtomicU64::new(0);

fn next_molecule_id() -> u64 {
    MOLECULE_ID_COUNTER.fetch_add(1, Ordering::Relaxed) + 1
}

// ---------------------------------------------------------------------------
// Molecule
// ---------------------------------------------------------------------------

/// A molecule: a collection of bonded atoms.
#[derive(Debug, Clone)]
pub struct Molecule {
    /// Atom IDs that form this molecule
    pub atom_ids: Vec<u64>,
    pub name: String,
    pub formula: String,
    pub molecule_id: u64,
    pub energy: f64,
    pub position: [f64; 3],
    pub is_organic: bool,
    pub functional_groups: Vec<String>,
    pub molecular_weight: f64,
}

impl Molecule {
    pub fn new(name: &str, atom_ids: Vec<u64>, position: [f64; 3], mw: f64) -> Self {
        Self {
            atom_ids,
            name: name.to_string(),
            formula: String::new(),
            molecule_id: next_molecule_id(),
            energy: 0.0,
            position,
            is_organic: false,
            functional_groups: Vec::new(),
            molecular_weight: mw,
        }
    }

    pub fn with_formula(mut self, formula: &str) -> Self {
        self.formula = formula.to_string();
        self
    }

    pub fn with_organic(mut self, organic: bool) -> Self {
        self.is_organic = organic;
        self
    }

    pub fn with_functional_groups(mut self, groups: Vec<&str>) -> Self {
        self.functional_groups = groups.into_iter().map(String::from).collect();
        self
    }

    pub fn atom_count(&self) -> usize {
        self.atom_ids.len()
    }
}

// ---------------------------------------------------------------------------
// ChemicalSystem
// ---------------------------------------------------------------------------

/// Manages molecular assembly and chemical reactions.
pub struct ChemicalSystem {
    pub molecules: Vec<Molecule>,
    pub reactions_occurred: u64,
    pub water_count: u64,
    pub amino_acid_count: u64,
    pub nucleotide_count: u64,
}

impl ChemicalSystem {
    pub fn new() -> Self {
        Self {
            molecules: Vec::new(),
            reactions_occurred: 0,
            water_count: 0,
            amino_acid_count: 0,
            nucleotide_count: 0,
        }
    }

    /// Form water molecules: 2H + O -> H2O.
    /// Returns the number of water molecules formed.
    pub fn form_water(&mut self, atoms: &mut Vec<Atom>) -> usize {
        let mut count = 0;

        loop {
            // Find two unbonded H and one O with fewer than 2 bonds
            let h_indices: Vec<usize> = atoms
                .iter()
                .enumerate()
                .filter(|(_, a)| a.atomic_number == 1 && a.bonds.is_empty())
                .map(|(i, _)| i)
                .collect();
            let o_idx = atoms
                .iter()
                .position(|a| a.atomic_number == 8 && a.bonds.len() < 2);

            if h_indices.len() < 2 || o_idx.is_none() {
                break;
            }

            let oi = o_idx.unwrap();
            let h1i = h_indices[0];
            let h2i = h_indices[1];

            let pos = atoms[oi].position;
            let o_id = atoms[oi].atom_id;
            let h1_id = atoms[h1i].atom_id;
            let h2_id = atoms[h2i].atom_id;

            // Form bonds
            atoms[h1i].bonds.push(o_id);
            atoms[h2i].bonds.push(o_id);
            atoms[oi].bonds.push(h1_id);
            atoms[oi].bonds.push(h2_id);

            let water = Molecule::new("water", vec![h1_id, h2_id, o_id], pos, 18.0)
                .with_formula("H2O");
            self.molecules.push(water);
            self.water_count += 1;
            count += 1;
        }

        count
    }

    /// Form methane: C + 4H -> CH4.
    pub fn form_methane(&mut self, atoms: &mut Vec<Atom>) -> usize {
        let mut count = 0;

        loop {
            let c_idx = atoms
                .iter()
                .position(|a| a.atomic_number == 6 && a.bonds.is_empty());
            let h_indices: Vec<usize> = atoms
                .iter()
                .enumerate()
                .filter(|(_, a)| a.atomic_number == 1 && a.bonds.is_empty())
                .map(|(i, _)| i)
                .collect();

            if c_idx.is_none() || h_indices.len() < 4 {
                break;
            }

            let ci = c_idx.unwrap();
            let his: Vec<usize> = h_indices[..4].to_vec();

            let pos = atoms[ci].position;
            let c_id = atoms[ci].atom_id;
            let mut atom_ids = vec![c_id];

            for &hi in &his {
                let h_id = atoms[hi].atom_id;
                atoms[hi].bonds.push(c_id);
                atoms[ci].bonds.push(h_id);
                atom_ids.push(h_id);
            }

            let methane = Molecule::new("methane", atom_ids, pos, 16.0)
                .with_formula("CH4");
            self.molecules.push(methane);
            count += 1;
        }

        count
    }

    /// Form ammonia: N + 3H -> NH3.
    pub fn form_ammonia(&mut self, atoms: &mut Vec<Atom>) -> usize {
        let mut count = 0;

        loop {
            let n_idx = atoms
                .iter()
                .position(|a| a.atomic_number == 7 && a.bonds.is_empty());
            let h_indices: Vec<usize> = atoms
                .iter()
                .enumerate()
                .filter(|(_, a)| a.atomic_number == 1 && a.bonds.is_empty())
                .map(|(i, _)| i)
                .collect();

            if n_idx.is_none() || h_indices.len() < 3 {
                break;
            }

            let ni = n_idx.unwrap();
            let his: Vec<usize> = h_indices[..3].to_vec();

            let pos = atoms[ni].position;
            let n_id = atoms[ni].atom_id;
            let mut atom_ids = vec![n_id];

            for &hi in &his {
                let h_id = atoms[hi].atom_id;
                atoms[hi].bonds.push(n_id);
                atoms[ni].bonds.push(h_id);
                atom_ids.push(h_id);
            }

            let ammonia = Molecule::new("ammonia", atom_ids, pos, 17.0)
                .with_formula("NH3");
            self.molecules.push(ammonia);
            count += 1;
        }

        count
    }

    /// Form an amino acid from available atoms.
    /// Simplified: requires 2C + 5H + 2O + 1N minimum (glycine).
    pub fn form_amino_acid(&mut self, atoms: &mut Vec<Atom>, aa_type: &str) -> bool {
        let carbons: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 6 && a.bonds.is_empty())
            .map(|(i, _)| i)
            .collect();
        let hydrogens: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 1 && a.bonds.is_empty())
            .map(|(i, _)| i)
            .collect();
        let oxygens: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 8 && a.bonds.len() < 2)
            .map(|(i, _)| i)
            .collect();
        let nitrogens: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 7 && a.bonds.is_empty())
            .map(|(i, _)| i)
            .collect();

        if carbons.len() < 2 || hydrogens.len() < 5 || oxygens.len() < 2 || nitrogens.is_empty() {
            return false;
        }

        let pos = atoms[carbons[0]].position;
        let anchor_id = atoms[carbons[0]].atom_id;

        // Collect atom IDs and form bonds
        let mut atom_ids = Vec::new();
        let all_indices: Vec<usize> = carbons[..2]
            .iter()
            .chain(hydrogens[..5].iter())
            .chain(oxygens[..2].iter())
            .chain(nitrogens[..1].iter())
            .copied()
            .collect();

        for &idx in &all_indices {
            let aid = atoms[idx].atom_id;
            atom_ids.push(aid);
            if aid != anchor_id {
                atoms[idx].bonds.push(anchor_id);
                atoms[carbons[0]].bonds.push(aid);
            }
        }

        let aa = Molecule::new(aa_type, atom_ids, pos, 75.0)
            .with_organic(true)
            .with_functional_groups(vec!["amino", "carboxyl"]);
        self.molecules.push(aa);
        self.amino_acid_count += 1;
        true
    }

    /// Form a nucleotide (sugar + phosphate + base).
    /// Simplified: requires C5 + H8 + O4 + N2 minimum.
    pub fn form_nucleotide(&mut self, atoms: &mut Vec<Atom>, base: &str) -> bool {
        let carbons: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 6 && a.bonds.is_empty())
            .map(|(i, _)| i)
            .collect();
        let hydrogens: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 1 && a.bonds.is_empty())
            .map(|(i, _)| i)
            .collect();
        let oxygens: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 8 && a.bonds.len() < 2)
            .map(|(i, _)| i)
            .collect();
        let nitrogens: Vec<usize> = atoms
            .iter()
            .enumerate()
            .filter(|(_, a)| a.atomic_number == 7 && a.bonds.is_empty())
            .map(|(i, _)| i)
            .collect();

        if carbons.len() < 5 || hydrogens.len() < 8 || oxygens.len() < 4 || nitrogens.len() < 2 {
            return false;
        }

        let pos = atoms[carbons[0]].position;
        let anchor_id = atoms[carbons[0]].atom_id;

        let mut atom_ids = Vec::new();
        let all_indices: Vec<usize> = carbons[..5]
            .iter()
            .chain(hydrogens[..8].iter())
            .chain(oxygens[..4].iter())
            .chain(nitrogens[..2].iter())
            .copied()
            .collect();

        for &idx in &all_indices {
            let aid = atoms[idx].atom_id;
            atom_ids.push(aid);
            if aid != anchor_id {
                atoms[idx].bonds.push(anchor_id);
                atoms[carbons[0]].bonds.push(aid);
            }
        }

        let name = format!("nucleotide-{}", base);
        let nuc = Molecule::new(&name, atom_ids, pos, 330.0)
            .with_organic(true)
            .with_functional_groups(vec!["sugar", "phosphate", "base"]);
        self.molecules.push(nuc);
        self.nucleotide_count += 1;
        true
    }

    /// Run catalyzed reactions to form complex molecules.
    pub fn catalyzed_reaction(
        &mut self,
        atoms: &mut Vec<Atom>,
        temperature: f64,
        catalyst_present: bool,
    ) -> u64 {
        let mut formed: u64 = 0;
        let ea_factor = if catalyst_present { 0.3 } else { 1.0 };
        let thermal = K_B * temperature;
        let mut rng = rand::thread_rng();

        // Try to form amino acids
        if thermal > 0.0 && atoms.len() > 10 {
            let aa_prob = (-5.0 * ea_factor / (thermal + 1e-20)).exp();
            if rng.gen::<f64>() < aa_prob {
                let idx = rng.gen_range(0..AMINO_ACIDS.len());
                let aa_type = AMINO_ACIDS[idx];
                if self.form_amino_acid(atoms, aa_type) {
                    formed += 1;
                    self.reactions_occurred += 1;
                }
            }
        }

        // Try to form nucleotides
        if thermal > 0.0 && atoms.len() > 19 {
            let nuc_prob = (-8.0 * ea_factor / (thermal + 1e-20)).exp();
            if rng.gen::<f64>() < nuc_prob {
                let bases = ['A', 'T', 'G', 'C'];
                let base = bases[rng.gen_range(0..4)];
                if self.form_nucleotide(atoms, &base.to_string()) {
                    formed += 1;
                    self.reactions_occurred += 1;
                }
            }
        }

        formed
    }

    /// Count molecules by type.
    pub fn molecule_census(&self) -> HashMap<String, usize> {
        let mut counts = HashMap::new();
        for m in &self.molecules {
            let key = if m.name.is_empty() {
                m.formula.clone()
            } else {
                m.name.clone()
            };
            *counts.entry(key).or_insert(0) += 1;
        }
        counts
    }

    pub fn to_compact(&self) -> String {
        let counts = self.molecule_census();
        let mut parts: Vec<String> = counts
            .iter()
            .map(|(k, v)| format!("{}:{}", k, v))
            .collect();
        parts.sort();
        format!(
            "CS[mol={} H2O={} aa={} nuc={} rxn={} {}]",
            self.molecules.len(),
            self.water_count,
            self.amino_acid_count,
            self.nucleotide_count,
            self.reactions_occurred,
            parts.join(",")
        )
    }
}
